### PR TITLE
scripts: extract_dts_includes: Fix extract_cells for a list

### PR DESCRIPTION
**Pull request by galak**
If we had something like:

gpios = <&gpiod 13 GPIO_INT_ACTIVE_LOW
         &gpiod 14 GPIO_INT_ACTIVE_LOW>;

The script blows up in extract_cells.  While extract_cells attempted to
handle more than a single item in such a list, it didn't manipulate the
prop list for the recursive calls properly.

Since we pop off items from the prop list as we use them we can easily
pass the prop list to the rescurive call to fix things.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

**Review by agross-korg — APPROVED** [review]
this looks reasonable

**Review by agross-oss — APPROVED** [review]

**Comment by nashif** [comment]
recheck

File: scripts/dts/extract_dts_includes.py [hunk 1]
```diff
@@ -228,16 +228,11 @@ def extract_controller(node_address, y_key, prefix, defs, def_label):
         insert_defs(node_address, defs, prop_def, prop_alias)
 
 
-def extract_cells(node_address, yaml, y_key, names, index, prefix, defs,
+def extract_cells(node_address, props, yaml, y_key, names, index, prefix, defs,
```
Comment (erwango) [inline scripts/dts/extract_dts_includes.py]: y_key is not needed anymore, I think

**Review by erwango — DISMISSED** [review]
I had a look on what is generated when using this, and there are a couple things to be aware/fix.

I used following dummy example:
```
	leds {
		compatible = "gpio-leds";
		green_led_2: led@0 {
			gpios = <&gpioa 5 GPIO_INT_ACTIVE_HIGH
				 &gpiob 4 GPIO_INT_ACTIVE_HIGH>;
			label = "User LD2";
		};
	};
	aliases {
		led0 = &green_led_2;
	};
```

```
/* led@0 */
#define GPIO_LEDS_0_GPIO_CONTROLLER	"GPIOA"
#define GPIO_LEDS_0_GPIO_CONTROLLER_0	"GPIOA"
#define GPIO_LEDS_0_GPIO_CONTROLLER_1	"GPIOB"
#define GPIO_LEDS_0_GPIO_FLAGS_0	4
#define GPIO_LEDS_0_GPIO_FLAGS_1	4
#define GPIO_LEDS_0_GPIO_PIN_0		5
#define GPIO_LEDS_0_GPIO_PIN_1		4
#define GPIO_LEDS_0_LABEL		"User LD2"
#define GPIO_LEDS_0_GPIO_FLAGS		GPIO_LEDS_0_GPIO_FLAGS_0
#define GPIO_LEDS_0_GPIO_PIN		GPIO_LEDS_0_GPIO_PIN_0
#define LED0_GPIO_CONTROLLER_0		GPIO_LEDS_0_GPIO_CONTROLLER_0
#define LED0_GPIO_CONTROLLER_1		GPIO_LEDS_0_GPIO_CONTROLLER_1
#define LED0_GPIO_FLAGS			GPIO_LEDS_0_GPIO_FLAGS
#define LED0_GPIO_PIN			GPIO_LEDS_0_GPIO_PIN
#define LED0_LABEL			GPIO_LEDS_0_LABEL
```

So one should know that `GPIO_LEDS_0_GPIO_CONTROLLER, GPIO_LEDS_0_GPIO_FLAGS and GPIO_LEDS_0_GPIO_PIN` will refer to first item. Not sure there is something to fix though, as _0 and _1 are correctly generated. 

What I think should be fixed is that `LED0_GPIO_FLAGS and LED0_GPIO_PIN` are not generated for second item (while we get the _0 and _1 versions for _CONTROLLER)

**Comment by galak** [comment]
@erwango do you mind taking this over and making the fixes you mentioned.

File: scripts/dts/extract_dts_includes.py [hunk 2]
```diff
@@ -443,7 +438,11 @@ def extract_property(node_compat, yaml, node_address, y_key, y_val, names,
                         names[p_index], p_index, defs, def_label)
     elif 'clocks' in y_key or 'gpios' in y_key:
         extract_controller(node_address, y_key, prefix, defs, def_label)
-        extract_cells(node_address, yaml, y_key,
+        try:
+            props = list(reduced[node_address]['props'].get(y_key))
```
[the remark this answers is not in the document]
Comment (galak) [reply scripts/dts/extract_dts_includes.py]: I was being a bit lazy in my approach, was wondering if keeping extract_cells and add a second function that extract_cells calls.

**Comment by erwango** [comment]
> @erwango do you mind taking this over and making the fixes you mentioned.

Sure, I could do it, is there a real life example to work on?

**Comment by galak** [comment]
> Sure, I could do it, is there a real life example to work on?

The case you have is pretty much similar to my real life one, the only difference is its not for led.

**Comment by erwango** [comment]
> The case you have is pretty much similar to my real life one, the only difference is its not for led.

One point to check: do we want node to be a gpio-device?
For instance (which is a bit different case, but I think we should start from that):
```
&spi3 {
	status = "ok";

	spbtle-rf@0 {
		compatible = "st,spbtle-rf";
		reg = <0>;
		spi-max-frequency = <8000000>;
		label = "SPBTLE-RF";
		cs-gpios = <&gpiod 13 GPIO_INT_ACTIVE_LOW>;
		reset-gpios = <&gpioa 8 GPIO_INT_ACTIVE_LOW>;
		irq-gpios = <&gpioe 6 GPIO_INT_ACTIVE_LOW>;
		};
};
```

This node is spi-device. We can add cs-gpios, reset-gpios and irq-gpios in its yaml properties,
and with minor script modification (take property name into account in define label generation), we can get things generated.
For instance (not fully working yet, will be easier tomorrow  with fresh mind)
```
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_BASE_ADDRESS_0		0x0
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_BUS_NAME			"SPI_3"
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_CS_GPIOS_CONTROLLER		"GPIOD"
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_GPIO_FLAGS_0			0
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_GPIO_PIN_0			6
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_IRQ_GPIOS_CONTROLLER		"GPIOE"
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_LABEL				"SPBTLE-RF"
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_RESET_GPIOS_CONTROLLER		"GPIOA"
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_SPI_MAX_FREQUENCY		8000000
```

Would you be happy with that (ie not using gpio-device for this kind of case)?
Then, I'll have a look to multiple pins.

**Comment by erwango** [comment]
@galak:

Unlike mentioned in the previous commit, it might be clearer to handle things separately.
So my proposal is the following.
If we have, in led0, something like:
```
gpios = <&gpiod 13 GPIO_INT_ACTIVE_LOW
&gpiod 14 GPIO_INT_ACTIVE_LOW>;
```
generate something like:
```
/* led@0 */
#define GPIO_LEDS_0_GPIO_CONTROLLER_0	"GPIOA"
#define GPIO_LEDS_0_GPIO_CONTROLLER_1	"GPIOB"
#define GPIO_LEDS_0_GPIO_FLAGS_0	4
#define GPIO_LEDS_0_GPIO_FLAGS_1	4
#define GPIO_LEDS_0_GPIO_PIN_0		5
#define GPIO_LEDS_0_GPIO_PIN_1		4
#define GPIO_LEDS_0_LABEL		"User LD2"
#define LED0_GPIO_CONTROLLER_0		GPIO_LEDS_0_GPIO_CONTROLLER_0
#define LED0_GPIO_CONTROLLER_1		GPIO_LEDS_0_GPIO_CONTROLLER_1
#define LED0_GPIO_FLAGS_0			GPIO_LEDS_0_GPIO_FLAGS_0
#define LED0_GPIO_FLAGS_1			GPIO_LEDS_0_GPIO_FLAGS_1
#define LED0_GPIO_PIN_0			GPIO_LEDS_0_GPIO_PIN_0
#define LED0_GPIO_PIN_1			GPIO_LEDS_0_GPIO_PIN_1
#define LED0_LABEL			GPIO_LEDS_0_LABEL
```
Which means:
-only the duplicate objects have an index (gpio_controller, pins and flags in this case)
-aliases indexes are generated on the same basis (ie no more alias on the item _0)

According to @agross-korg aliases to _0 index were implemented for a purpose which isn't in use currently, so this shouldn't cause any problem.
Are you fine with that?

**Comment by erwango** [comment]
@galak , I've pushed a new version of the PR according to my last comment.

**Review by GAnthony — APPROVED** [review]
LGTM

**Pull request merged**